jax/lax/lax_control_flow.py

```diff
@@ -192,17 +192,24 @@ def fori_loop(lower, upper, body_fun, init_val):
            "got {} and {}")
     raise TypeError(msg.format(lower_dtype.name, upper_dtype.name))
 
-  # If we can specialize on the trip count, call scan instead of a while_loop
-  # to enable efficient reverse-mode differentiation.
   try:
     lower_ = int(lower)
     upper_ = int(upper)
   except TypeError:
-    use_scan = False
+    static_trip_count = False
   else:
-    use_scan = False  # TODO(mattjj): re-enable this
-
-  if use_scan:
+    static_trip_count = True
+
+  if static_trip_count and lower_ == upper_:
```

Comment:
I am not sure about this change. The more special cases we add in the code, the more complex the mental model our users need to understand. I personally would prefer to not treat this as a special case and stick to the model that for/while loops are traced by JAX, independent on the number of iterations.

Reply:
I hear you, at least in the abstract. But we've seen two cases now where this would be a convenient behavior (one from Jake and one from Peter). What's an example where this change would be confusing?

Reply:
Without at least one concrete example in which this leads to confusion to counter our two examples where this behavior is desirable, I'm likely to merge this.

Reply:
This is not a big deal, and I am fine it we merge it. 

I am thinking of something along the lines of code in #3285 `jax.lax.fori_loop(0, len(vals), body_fun, init)` that works when passed an empty `vals` and breaks if you pass a non-empty `vals`. It would be nice to not have to explain to a user why this is happening. 

I suspect this is a rare use case (transitioning from empty array to non-empty array).